pkg/install/install.go

```diff
@@ -0,0 +1,464 @@
+	info, err := os.Stat(dir)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return info.IsDir()
+}
+
+func fileExists(file string) bool {
+	info, err := os.Stat(file)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
+}
+
+func mkdir(path string) {
+	if err := os.MkdirAll(path, 0777); err != nil {
+		logrus.WithError(err).Fatalf("Failed to create directory %s", path)
+	}
+
+}
+
+func loadConfig() config {
+	var c config
+	err := envconfig.Process("", &c)
+	if err != nil {
+		logrus.Fatal(err.Error())
+	}
+
+	if c.TLSAssetsDir == "" {
+		c.TLSAssetsDir = "/calico-secrets"
+	}
+
+	return c
+}
+
+func Install() error {
+	// Clean up any existing binaries / config / assets.
+	if err := os.Remove("/host/opt/cni/bin/calico"); err != nil {
+		logrus.WithError(err).Warnf("Error removing old plugin")
+	}
+	if err := os.Remove("/host/opt/cni/bin/calico-ipam"); err != nil {
+		logrus.WithError(err).Warnf("Error removing old IPAM plugin")
+	}
+	if err := os.RemoveAll("/host/etc/cni/net.d/calico-tls"); err != nil {
+		logrus.WithError(err).Warnf("Error removing old TLS directory")
+	}
+
+	// Load config.
+	c := loadConfig()
+
+	// Determine if we're running as a Kubernetes pod.
+	var kubecfg *rest.Config
+
+	serviceAccountTokenFile := "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	c.ServiceAccountToken = make([]byte, 0)
+	var err error
+	if fileExists(serviceAccountTokenFile) {
+		log.Info("Running as a Kubernetes pod")
+		kubecfg, err = rest.InClusterConfig()
+		if err != nil {
+			return err
+		}
+		err = rest.LoadTLSFiles(kubecfg)
+		if err != nil {
+			return err
+		}
+
+		c.ServiceAccountToken, err = ioutil.ReadFile(serviceAccountTokenFile)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Copy over any TLS assets from the SECRETS_MOUNT_DIR to the host.
+	// First check if the dir exists and has anything in it.
+	if directoryExists(c.TLSAssetsDir) {
+		logrus.Info("Installing any TLS assets")
+		mkdir("/host/etc/cni/net.d/calico-tls")
+		if err := copyFile(fmt.Sprintf("%s/%s", c.TLSAssetsDir, "etcd-ca"), "/host/etc/cni/net.d/calico-tls/etcd-ca"); err != nil {
+			logrus.Warnf("Missing etcd-ca")
+		}
+		if err := copyFile(fmt.Sprintf("%s/%s", c.TLSAssetsDir, "etcd-cert"), "/host/etc/cni/net.d/calico-tls/etcd-cert"); err != nil {
+			logrus.Warnf("Missing etcd-cert")
+		}
+		if err := copyFile(fmt.Sprintf("%s/%s", c.TLSAssetsDir, "etcd-key"), "/host/etc/cni/net.d/calico-tls/etcd-key"); err != nil {
+			logrus.Warnf("Missing etcd-key")
+		}
+	}
+
+	// Place the new binaries if the directory is writeable.
+	dirs := []string{"/host/opt/cni/bin", "/host/secondary-bin-dir"}
+	for _, d := range dirs {
+		if unix.Access(d, unix.W_OK) != nil {
+			logrus.Infof("%s is not writeable, skipping", d)
+			continue
+		}
+
+		// Iterate through each binary we might want to install.
+		// TODO: Include all the binaries we ship.
+		binaries := []string{"calico", "calico-ipam", "bandwidth", "flannel", "loopback", "host-local", "portmap", "tuning"}
```

Comment:
This deviates from the original script - in that the script just lists all files in the bin directory.  Any particular reason we don't do that here?

Reply:
Changed to go over all files in  /opt/cni/bin/